File: crates/crf-protocol/src/wire.rs
use std::{collections::BTreeSet, fmt};

use serde::{Deserialize, Serialize};

use crate::{ExecutionBackend, NodeSnapshot, Resources, valid_identifier};

pub const PROTOCOL_VERSION: u16 = 1;
pub const MAX_WIRE_MESSAGE_BYTES: usize = 256 * 1024;
pub const MAX_COMMAND_TTL_MS: u64 = 5 * 60 * 1000;
pub const MAX_JIT_CONFIG_BYTES: usize = 64 * 1024;
pub const MAX_ACTIVE_PLACEMENTS: usize = 1024;

#[derive(Clone, Eq, PartialEq, Serialize, Deserialize)]
#[serde(transparent)]
pub struct SecretString(String);

impl SecretString {
    pub fn new(value: impl Into<String>) -> Result<Self, WireError> {
        let value = value.into();
        if !valid_jit_config(&value) {
            return Err(WireError::InvalidSecret);
        }
        Ok(Self(value))
    }

    pub fn expose_secret(&self) -> &str {
        &self.0
    }
}

impl fmt::Debug for SecretString {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str("SecretString([REDACTED])")
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct NodeEnvelope {
    pub protocol_version: u16,
    pub message_id: String,
    pub node_id: String,
    pub node_generation: u64,
    pub sent_at_unix_ms: u64,
    pub payload: NodeMessage,
}

impl NodeEnvelope {
    pub fn validate(&self) -> Result<(), WireError> {
        if self.protocol_version != PROTOCOL_VERSION {
            return Err(WireError::UnsupportedProtocolVersion);
        }
        if !valid_identifier(&self.message_id) {
            return Err(WireError::InvalidMessageId);
        }
        if !valid_identifier(&self.node_id) {
            return Err(WireError::InvalidNodeId);
        }
        if self.node_generation == 0 || self.sent_at_unix_ms == 0 {
            return Err(WireError::InvalidEnvelope);
        }
        self.payload.validate(&self.node_id, self.node_generation)
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case", deny_unknown_fields)]
pub enum NodeMessage {
    Register {
        node: NodeSnapshot,
        agent_version: String,
    },
    Heartbeat {
        available: Resources,
        active_placements: BTreeSet<String>,
    },
    CommandAck {
        command_id: String,
        idempotency_key: String,
        status: CommandAckStatus,
        detail_code: Option<String>,
    },
    PlacementUpdate {
        placement_id: String,
        command_id: String,
        state: PlacementState,
        detail_code: Option<String>,
    },
}

impl NodeMessage {
    fn validate(&self, node_id: &str, node_generation: u64) -> Result<(), WireError> {
        match self {
            Self::Register {
                node,
                agent_version,
            } => {
                node.validate()
                    .map_err(|_| WireError::InvalidNodeSnapshot)?;
                if node.node_id != node_id || node.generation != node_generation {
                    return Err(WireError::IdentityMismatch);
                }
                if !valid_version(agent_version) {
                    return Err(WireError::InvalidAgentVersion);
                }
            }
            Self::Heartbeat {
                available: _,
                active_placements,
            } => {
                if active_placements.len() > MAX_ACTIVE_PLACEMENTS
                    || active_placements
                        .iter()
                        .any(|placement| !valid_identifier(placement))
                {
                    return Err(WireError::InvalidPlacementId);
                }
            }
            Self::CommandAck {
                command_id,
                idempotency_key,
                detail_code,
                ..
            } => {
                validate_command_identity(command_id, idempotency_key)?;
                validate_detail_code(detail_code)?;
            }
            Self::PlacementUpdate {
                placement_id,
                command_id,
                detail_code,
                ..
            } => {
                if !valid_identifier(placement_id) || !valid_identifier(command_id) {
                    return Err(WireError::InvalidPlacementId);
                }
                validate_detail_code(detail_code)?;
            }
        }
        Ok(())
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct NodeMessageResponse {
    pub protocol_version: u16,
    pub message_id: String,
    pub status: MessageResponseStatus,
    pub code: Option<String>,
    pub command: Option<ControllerEnvelope>,
}

impl NodeMessageResponse {
    pub fn validate(&self, now_unix_ms: u64) -> Result<(), WireError> {
        if self.protocol_version != PROTOCOL_VERSION {
            return Err(WireError::UnsupportedProtocolVersion);
        }
        if !valid_identifier(&self.message_id) {
            return Err(WireError::InvalidMessageId);
        }
        validate_detail_code(&self.code)?;
        match self.status {
            MessageResponseStatus::Rejected if self.code.is_none() || self.command.is_some() => {
                return Err(WireError::InvalidResponse);
            }
            MessageResponseStatus::Accepted | MessageResponseStatus::Duplicate
                if self.code.is_some() =>
            {
                return Err(WireError::InvalidResponse);
            }
            _ => {}
        }
        if let Some(command) = &self.command {
            command.validate(now_unix_ms)?;
        }
        Ok(())
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MessageResponseStatus {
    Accepted,
    Duplicate,
    Rejected,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CommandAckStatus {
    Accepted,
    Duplicate,
    Rejected,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PlacementState {
    Accepted,
    Starting,
    Observed,
    Running,
    Finished,
    Failed,
    Cancelled,
}

impl PlacementState {
    pub const fn terminal(self) -> bool {
        matches!(self, Self::Finished | Self::Failed | Self::Cancelled)
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ControllerEnvelope {
    pub protocol_version: u16,
    pub command_id: String,
    pub idempotency_key: String,
    pub node_id: String,
    pub node_generation: u64,
    pub issued_at_unix_ms: u64,
    pub expires_at_unix_ms: u64,
    pub payload: ControllerCommand,
}

impl ControllerEnvelope {
    pub fn validate(&self, now_unix_ms: u64) -> Result<(), WireError> {
        if self.protocol_version != PROTOCOL_VERSION {
            return Err(WireError::UnsupportedProtocolVersion);
        }
        validate_command_identity(&self.command_id, &self.idempotency_key)?;
        if !valid_identifier(&self.node_id) || self.node_generation == 0 {
            return Err(WireError::InvalidNodeId);
        }
        if self.issued_at_unix_ms == 0
            || self.expires_at_unix_ms < self.issued_at_unix_ms
            || self.expires_at_unix_ms - self.issued_at_unix_ms > MAX_COMMAND_TTL_MS
        {
            return Err(WireError::InvalidCommandLifetime);
        }
        if now_unix_ms > self.expires_at_unix_ms {
            return Err(WireError::ExpiredCommand);
        }
        self.payload.validate()
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case", deny_unknown_fields)]
pub enum ControllerCommand {
    StartPlacement {
        placement_id: String,
        work_id: String,
        pool_id: String,
        runner_name: String,
        resources: Resources,
        execution_backend: ExecutionBackend,
        jit_config: SecretString,
    },
    CancelPlacement {
        placement_id: String,
    },
    SetDrain {
        draining: bool,
    },
}

impl ControllerCommand {
    fn validate(&self) -> Result<(), WireError> {
        match self {
            Self::StartPlacement {
                placement_id,
                work_id,
                pool_id,
                runner_name,
                resources,
                jit_config,
                ..
            } => {
                if [placement_id, work_id, pool_id, runner_name]
                    .into_iter()
                    .any(|value| !valid_identifier(value))
                {
                    return Err(WireError::InvalidPlacementId);
                }
                if resources.cpu_millis == 0 || resources.memory_bytes == 0 {
                    return Err(WireError::InvalidResources);
                }
                if !valid_jit_config(jit_config.expose_secret()) {
                    return Err(WireError::InvalidSecret);
                }
            }
            Self::CancelPlacement { placement_id } => {
                if !valid_identifier(placement_id) {
                    return Err(WireError::InvalidPlacementId);
                }
            }
            Self::SetDrain { .. } => {}
        }
        Ok(())
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum WireError {
    TooLarge,
    InvalidJson,
    InvalidFrame,
    InvalidResponse,
    UnsupportedProtocolVersion,
    InvalidEnvelope,
    InvalidMessageId,
    InvalidNodeId,
    InvalidNodeSnapshot,
    InvalidAgentVersion,
    InvalidResources,
    InvalidPlacementId,
    InvalidCommandId,
    InvalidIdempotencyKey,
    InvalidCommandLifetime,
    ExpiredCommand,
    IdentityMismatch,
    InvalidDetailCode,
    InvalidSecret,
}

pub fn encode_frame(payload: &[u8]) -> Result<Vec<u8>, WireError> {
    if payload.is_empty() || payload.len() > MAX_WIRE_MESSAGE_BYTES {
        return Err(WireError::TooLarge);
    }
    let length = u32::try_from(payload.len()).map_err(|_| WireError::TooLarge)?;
    let mut frame = Vec::with_capacity(4 + payload.len());
    frame.extend_from_slice(&length.to_be_bytes());
    frame.extend_from_slice(payload);
    Ok(frame)
}

pub fn decode_frame(frame: &[u8]) -> Result<&[u8], WireError> {
    if frame.len() < 4 {
        return Err(WireError::InvalidFrame);
    }
    let length =
        u32::from_be_bytes(frame[..4].try_into().map_err(|_| WireError::InvalidFrame)?) as usize;
    if length == 0 || length > MAX_WIRE_MESSAGE_BYTES || frame.len() != length + 4 {
        return Err(WireError::InvalidFrame);
    }
    Ok(&frame[4..])
}

pub fn encode_node_message(message: &NodeEnvelope) -> Result<Vec<u8>, WireError> {
    message.validate()?;
    encode(message)
}

pub fn decode_node_message(bytes: &[u8]) -> Result<NodeEnvelope, WireError> {
    let message: NodeEnvelope = decode(bytes)?;
    message.validate()?;
    Ok(message)
}

pub fn encode_node_response(
    response: &NodeMessageResponse,
    now_unix_ms: u64,
) -> Result<Vec<u8>, WireError> {
    response.validate(now_unix_ms)?;
    encode(response)
}

pub fn decode_node_response(
    bytes: &[u8],
    now_unix_ms: u64,
) -> Result<NodeMessageResponse, WireError> {
    let response: NodeMessageResponse = decode(bytes)?;
    response.validate(now_unix_ms)?;
    Ok(response)
}

pub fn encode_controller_command(
    command: &ControllerEnvelope,
    now_unix_ms: u64,
) -> Result<Vec<u8>, WireError> {
    command.validate(now_unix_ms)?;
    encode(command)
}

pub fn decode_controller_command(
    bytes: &[u8],
    now_unix_ms: u64,
) -> Result<ControllerEnvelope, WireError> {
    let command: ControllerEnvelope = decode(bytes)?;
    command.validate(now_unix_ms)?;
    Ok(command)
}

fn encode<T: Serialize>(value: &T) -> Result<Vec<u8>, WireError> {
    let encoded = serde_json::to_vec(value).map_err(|_| WireError::InvalidJson)?;
    if encoded.len() > MAX_WIRE_MESSAGE_BYTES {
        return Err(WireError::TooLarge);
    }
    Ok(encoded)
}

fn decode<T: for<'de> Deserialize<'de>>(bytes: &[u8]) -> Result<T, WireError> {
    if bytes.is_empty() || bytes.len() > MAX_WIRE_MESSAGE_BYTES {
        return Err(WireError::TooLarge);
    }
    serde_json::from_slice(bytes).map_err(|_| WireError::InvalidJson)
}

fn validate_command_identity(command_id: &str, idempotency_key: &str) -> Result<(), WireError> {
    if !valid_identifier(command_id) {
        return Err(WireError::InvalidCommandId);
    }
    if !valid_identifier(idempotency_key) {
        return Err(WireError::InvalidIdempotencyKey);
    }
    Ok(())
}

fn validate_detail_code(detail_code: &Option<String>) -> Result<(), WireError> {
    if detail_code
        .as_deref()
        .is_some_and(|detail| !valid_identifier(detail))
    {
        return Err(WireError::InvalidDetailCode);
    }
    Ok(())
}

fn valid_version(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= 64
        && value
            .chars()
            .all(|ch| ch.is_ascii_alphanumeric() || matches!(ch, '.' | '-' | '+'))
}

fn valid_jit_config(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= MAX_JIT_CONFIG_BYTES
        && value.bytes().all(|byte| {
            byte.is_ascii_alphanumeric()
                || matches!(byte, b'.' | b'_' | b'+' | b'/' | b'=' | b':' | b'-')
        })
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeSet;

    use crate::{Architecture, OperatingSystem};

    use super::*;

    const GIB: u64 = 1024 * 1024 * 1024;

    fn node() -> NodeSnapshot {
        NodeSnapshot {
            node_id: "dookie".into(),
            generation: 7,
            os: OperatingSystem::Linux,
            arch: Architecture::X86_64,
            execution_backends: BTreeSet::from([ExecutionBackend::Container]),
            capabilities: BTreeSet::from(["github-actions".into(), "x64".into()]),
            total: Resources::new(20_000, 48 * GIB),
            available: Resources::new(18_000, 40 * GIB),
            draining: false,
        }
    }

    #[test]
    fn node_registration_round_trips_with_versioned_identity() {
        let envelope = NodeEnvelope {
            protocol_version: PROTOCOL_VERSION,
            message_id: "msg-register-1".into(),
            node_id: "dookie".into(),
            node_generation: 7,
            sent_at_unix_ms: 1_787_070_000_000,
            payload: NodeMessage::Register {
                node: node(),
                agent_version: "0.1.0".into(),
            },
        };
        let encoded = encode_node_message(&envelope).expect("registration encodes");
        assert_eq!(decode_node_message(&encoded), Ok(envelope));
    }

    #[test]
    fn registration_cannot_claim_a_different_authenticated_identity() {
        let envelope = NodeEnvelope {
            protocol_version: PROTOCOL_VERSION,
            message_id: "msg-register-1".into(),
            node_id: "steamy".into(),
            node_generation: 7,
            sent_at_unix_ms: 1_787_070_000_000,
            payload: NodeMessage::Register {
                node: node(),
                agent_version: "0.1.0".into(),
            },
        };
        assert_eq!(envelope.validate(), Err(WireError::IdentityMismatch));
    }

    #[test]
    fn command_debug_output_never_contains_jit_secret() {
        let secret = "super-secret-jit-config";
        let command = ControllerEnvelope {
            protocol_version: PROTOCOL_VERSION,
            command_id: "cmd-1".into(),
            idempotency_key: "idem-1".into(),
            node_id: "steamy".into(),
            node_generation: 3,
            issued_at_unix_ms: 1_787_070_000_000,
            expires_at_unix_ms: 1_787_070_030_000,
            payload: ControllerCommand::StartPlacement {
                placement_id: "placement-1".into(),
                work_id: "work-1".into(),
                pool_id: "windows".into(),
                runner_name: "crf-steamy-1".into(),
                resources: Resources::new(4_000, 8 * GIB),
                execution_backend: ExecutionBackend::NativeProcess,
                jit_config: SecretString::new(secret).expect("valid secret"),
            },
        };
        let debug = format!("{command:?}");
        assert!(!debug.contains(secret));
        assert!(debug.contains("[REDACTED]"));
    }

    #[test]
    fn expired_commands_fail_closed() {
        let command = ControllerEnvelope {
            protocol_version: PROTOCOL_VERSION,
            command_id: "cmd-1".into(),
            idempotency_key: "idem-1".into(),
            node_id: "dookie".into(),
            node_generation: 7,
            issued_at_unix_ms: 1_000,
            expires_at_unix_ms: 2_000,
            payload: ControllerCommand::SetDrain { draining: true },
        };
        assert_eq!(command.validate(2_001), Err(WireError::ExpiredCommand));
    }

    #[test]
    fn oversized_wire_messages_are_rejected_before_json_decode() {
        let bytes = vec![b' '; MAX_WIRE_MESSAGE_BYTES + 1];
        assert_eq!(decode_node_message(&bytes), Err(WireError::TooLarge));
    }

    #[test]
    fn length_prefixed_frames_round_trip() {
        let payload = br#"{"protocol_version":1}"#;
        let frame = encode_frame(payload).expect("frame encodes");
        assert_eq!(&frame[..4], &(payload.len() as u32).to_be_bytes());
        assert_eq!(decode_frame(&frame), Ok(payload.as_slice()));
    }

    #[test]
    fn malformed_or_oversized_frames_fail_before_payload_processing() {
        assert_eq!(decode_frame(&[0, 0, 0]), Err(WireError::InvalidFrame));
        assert_eq!(
            decode_frame(&[0, 0, 0, 5, b'a']),
            Err(WireError::InvalidFrame)
        );
        assert_eq!(encode_frame(&[]), Err(WireError::TooLarge));
        assert_eq!(
            encode_frame(&vec![b'x'; MAX_WIRE_MESSAGE_BYTES + 1]),
            Err(WireError::TooLarge)
        );
    }

    #[test]
    fn rust_wire_decoder_rejects_unknown_envelope_payload_and_nested_fields() {
        let envelope = NodeEnvelope {
            protocol_version: PROTOCOL_VERSION,
            message_id: "msg-strict-1".into(),
            node_id: "dookie".into(),
            node_generation: 7,
            sent_at_unix_ms: 1_787_070_000_000,
            payload: NodeMessage::Register {
                node: node(),
                agent_version: "0.1.0".into(),
            },
        };
        let encoded = encode_node_message(&envelope).expect("registration encodes");
        let value: serde_json::Value = serde_json::from_slice(&encoded).expect("valid json");

        let mut envelope_extra = value.clone();
        envelope_extra
            .as_object_mut()
            .expect("envelope object")
            .insert("surprise".into(), serde_json::Value::Bool(true));
        assert_eq!(
            decode_node_message(&serde_json::to_vec(&envelope_extra).expect("json")),
            Err(WireError::InvalidJson)
        );

        let mut payload_extra = value.clone();
        payload_extra["payload"]
            .as_object_mut()
            .expect("payload object")
            .insert("surprise".into(), serde_json::Value::Bool(true));
        assert_eq!(
            decode_node_message(&serde_json::to_vec(&payload_extra).expect("json")),
            Err(WireError::InvalidJson)
        );

        let mut node_extra = value;
        node_extra["payload"]["node"]
            .as_object_mut()
            .expect("node object")
            .insert("surprise".into(), serde_json::Value::Bool(true));
        assert_eq!(
            decode_node_message(&serde_json::to_vec(&node_extra).expect("json")),
            Err(WireError::InvalidJson)
        );
    }

    #[test]
    fn response_status_cannot_contradict_code_or_attached_command() {
        let command = decode_controller_command(
            include_bytes!("../../../tests/fixtures/distributed-controller-command-v1.json"),
            1_787_070_001_000,
        )
        .expect("fixture command");

        let rejected_with_command = NodeMessageResponse {
            protocol_version: PROTOCOL_VERSION,
            message_id: "message-1".into(),
            status: MessageResponseStatus::Rejected,
            code: Some("controller_rejected".into()),
            command: Some(command),
        };
        assert_eq!(
            rejected_with_command.validate(1_787_070_001_000),
            Err(WireError::InvalidResponse)
        );

        let accepted_with_code = NodeMessageResponse {
            protocol_version: PROTOCOL_VERSION,
            message_id: "message-2".into(),
            status: MessageResponseStatus::Accepted,
            code: Some("unexpected_code".into()),
            command: None,
        };
        assert_eq!(
            accepted_with_code.validate(1_787_070_001_000),
            Err(WireError::InvalidResponse)
        );

        let rejected_without_code = NodeMessageResponse {
            protocol_version: PROTOCOL_VERSION,
            message_id: "message-3".into(),
            status: MessageResponseStatus::Rejected,
            code: None,
            command: None,
        };
        assert_eq!(
            rejected_without_code.validate(1_787_070_001_000),
            Err(WireError::InvalidResponse)
        );
    }

    #[test]
    fn shared_elixir_controller_fixture_decodes_in_rust() {
        let bytes =
            include_bytes!("../../../tests/fixtures/distributed-controller-command-v1.json");
        let command = decode_controller_command(bytes, 1_787_070_001_000)
            .expect("shared controller command fixture decodes");

        assert_eq!(command.command_id, "command-1");
        assert_eq!(command.idempotency_key, "idempotency-1");
        assert_eq!(command.node_id, "steamy");
        assert_eq!(command.node_generation, 3);
        match command.payload {
            ControllerCommand::StartPlacement {
                placement_id,
                work_id,
                pool_id,
                runner_name,
                resources,
                execution_backend,
                jit_config,
            } => {
                assert_eq!(placement_id, "placement-1");
                assert_eq!(work_id, "work-1");
                assert_eq!(pool_id, "build");
                assert_eq!(runner_name, "crf-steamy-1");
                assert_eq!(resources, Resources::new(2_000, 4 * GIB));
                assert_eq!(execution_backend, ExecutionBackend::NativeProcess);
                assert_eq!(jit_config.expose_secret(), "jit-config-abc123==");
            }
            other => panic!("unexpected command payload: {other:?}"),
        }
    }
}

File: controller/lib/crf_controller/wire.ex
defmodule CrfController.Wire do
  alias CrfController.{Identifier, NodeCommand, PeerIdentity, Resources}

  @protocol_version 1
  @max_wire_message_bytes 256 * 1024
  @max_active_placements 1024
  @envelope_keys MapSet.new([
                   "protocol_version",
                   "message_id",
                   "node_id",
                   "node_generation",
                   "sent_at_unix_ms",
                   "payload"
                 ])

  def protocol_version, do: @protocol_version
  def max_wire_message_bytes, do: @max_wire_message_bytes

  @spec decode_node_envelope(binary(), PeerIdentity.t()) :: {:ok, map()} | {:error, atom()}
  def decode_node_envelope(binary, %PeerIdentity{} = peer)
      when is_binary(binary) and byte_size(binary) > 0 and
             byte_size(binary) <= @max_wire_message_bytes do
    with {:ok, envelope} <- decode_json(binary),
         :ok <- exact_keys(envelope, @envelope_keys),
         :ok <- protocol_version(envelope),
         {:ok, message_id} <- identifier(envelope["message_id"], :invalid_message_id),
         {:ok, node_id} <- identifier(envelope["node_id"], :invalid_node_id),
         :ok <- authenticated_identity(node_id, peer),
         {:ok, generation} <- positive_integer(envelope["node_generation"], :invalid_generation),
         {:ok, sent_at_unix_ms} <-
           positive_integer(envelope["sent_at_unix_ms"], :invalid_timestamp),
         {:ok, payload} <- parse_payload(envelope["payload"], node_id, generation) do
      {:ok,
       %{
         protocol_version: @protocol_version,
         message_id: message_id,
         node_id: node_id,
         node_generation: generation,
         sent_at_unix_ms: sent_at_unix_ms,
         payload: payload
       }}
    end
  end

  def decode_node_envelope(binary, %PeerIdentity{}) when is_binary(binary),
    do: {:error, :message_too_large}

  def decode_node_envelope(_, _), do: {:error, :invalid_transport_input}

  @spec encode_response(String.t(), :accepted | :duplicate | :rejected, atom() | nil) ::
          {:ok, binary()} | {:error, atom()}
  def encode_response(message_id, status, code \\ nil) do
    encode_response(message_id, status, code, nil, System.system_time(:millisecond))
  end

  @spec encode_response(
          String.t(),
          :accepted | :duplicate | :rejected,
          atom() | nil,
          NodeCommand.t() | nil,
          non_neg_integer()
        ) :: {:ok, binary()} | {:error, atom()}
  def encode_response(message_id, status, code, command, now_unix_ms) do
    with {:ok, message_id} <- identifier(message_id, :invalid_message_id),
         true <- status in [:accepted, :duplicate, :rejected],
         {:ok, code} <- response_code(code),
         {:ok, command} <- response_command(command, now_unix_ms) do
      json = %{
        "protocol_version" => @protocol_version,
        "message_id" => message_id,
        "status" => Atom.to_string(status),
        "code" => if(is_nil(code), do: :null, else: code),
        "command" => command
      }

      encoded = :json.encode(json) |> IO.iodata_to_binary()

      if byte_size(encoded) <= @max_wire_message_bytes,
        do: {:ok, encoded},
        else: {:error, :message_too_large}
    else
      false -> {:error, :invalid_response_status}
      {:error, reason} -> {:error, reason}
    end
  end

  defp parse_payload(%{"type" => "register"} = payload, node_id, generation) do
    with :ok <- exact_keys(payload, MapSet.new(["type", "node", "agent_version"])),
         {:ok, node} <- parse_node(payload["node"]),
         true <- node.id == node_id and node.generation == generation,
         {:ok, agent_version} <- agent_version(payload["agent_version"]) do
      {:ok, {:register, node, agent_version}}
    else
      false -> {:error, :identity_mismatch}
      {:error, reason} -> {:error, reason}
    end
  end

  defp parse_payload(%{"type" => "heartbeat"} = payload, _node_id, _generation) do
    with :ok <- exact_keys(payload, MapSet.new(["type", "available", "active_placements"])),
         {:ok, available} <- resources(payload["available"], false),
         {:ok, active} <- identifiers(payload["active_placements"], @max_active_placements) do
      {:ok, {:heartbeat, available, MapSet.new(active)}}
    end
  end

  defp parse_payload(%{"type" => "command_ack"} = payload, _node_id, _generation) do
    with :ok <-
           exact_keys(
             payload,
             MapSet.new(["type", "command_id", "idempotency_key", "status", "detail_code"])
           ),
         {:ok, command_id} <- identifier(payload["command_id"], :invalid_command_id),
         {:ok, idempotency_key} <-
           identifier(payload["idempotency_key"], :invalid_idempotency_key),
         {:ok, status} <- command_ack_status(payload["status"]),
         {:ok, detail_code} <- detail_code(payload["detail_code"]) do
      {:ok, {:command_ack, command_id, idempotency_key, status, detail_code}}
    end
  end

  defp parse_payload(%{"type" => "placement_update"} = payload, _node_id, _generation) do
    with :ok <-
           exact_keys(
             payload,
             MapSet.new(["type", "placement_id", "command_id", "state", "detail_code"])
           ),
         {:ok, placement_id} <- identifier(payload["placement_id"], :invalid_placement_id),
         {:ok, command_id} <- identifier(payload["command_id"], :invalid_command_id),
         {:ok, state} <- placement_state(payload["state"]),
         {:ok, detail_code} <- detail_code(payload["detail_code"]) do
      {:ok, {:placement_update, placement_id, command_id, state, detail_code}}
    end
  end

  defp parse_payload(_, _node_id, _generation), do: {:error, :invalid_payload}

  defp parse_node(node) when is_map(node) do
    keys =
      MapSet.new([
        "node_id",
        "generation",
        "os",
        "arch",
        "execution_backends",
        "capabilities",
        "total",
        "available",
        "draining"
      ])

    with :ok <- exact_keys(node, keys),
         {:ok, id} <- identifier(node["node_id"], :invalid_node_id),
         {:ok, generation} <- positive_integer(node["generation"], :invalid_generation),
         {:ok, os} <- operating_system(node["os"]),
         {:ok, arch} <- architecture(node["arch"]),
         {:ok, backends} <- execution_backends(node["execution_backends"]),
         {:ok, capabilities} <- identifiers(node["capabilities"], 256),
         {:ok, total} <- resources(node["total"], true),
         {:ok, available} <- resources(node["available"], false),
         true <- Resources.fits?(total, available),
         {:ok, draining} <- boolean(node["draining"], :invalid_draining) do
      {:ok,
       %{
         id: id,
         generation: generation,
         os: os,
         arch: arch,
         execution_backends: backends,
         capabilities: capabilities,
         total: total,
         available: available,
         draining: draining
       }}
    else
      false -> {:error, :available_exceeds_total}
      {:error, reason} -> {:error, reason}
    end
  end

  defp parse_node(_), do: {:error, :invalid_node}

  defp resources(%{"cpu_millis" => cpu, "memory_bytes" => memory} = value, positive?)
       when map_size(value) == 2 do
    with true <- is_integer(cpu) and cpu >= 0 and is_integer(memory) and memory >= 0,
         true <- not positive? or (cpu > 0 and memory > 0),
         {:ok, resources} <- Resources.new(%{cpu_millis: cpu, memory_bytes: memory}) do
      {:ok, resources}
    else
      _ -> {:error, :invalid_resources}
    end
  end

  defp resources(_, _), do: {:error, :invalid_resources}

  defp exact_keys(value, expected) when is_map(value) do
    if Map.keys(value) |> MapSet.new() |> MapSet.equal?(expected),
      do: :ok,
      else: {:error, :unexpected_fields}
  end

  defp protocol_version(%{"protocol_version" => @protocol_version}), do: :ok
  defp protocol_version(_), do: {:error, :unsupported_protocol_version}

  defp authenticated_identity(node_id, %PeerIdentity{node_id: node_id}), do: :ok
  defp authenticated_identity(_, _), do: {:error, :authenticated_identity_mismatch}

  defp identifier(value, error) do
    if Identifier.valid?(value), do: {:ok, value}, else: {:error, error}
  end

  defp identifiers(values, max) when is_list(values) and length(values) <= max do
    if Enum.all?(values, &Identifier.valid?/1) and
         MapSet.size(MapSet.new(values)) == length(values),
       do: {:ok, values},
       else: {:error, :invalid_identifier_list}
  end

  defp identifiers(_, _), do: {:error, :invalid_identifier_list}

  defp positive_integer(value, _error) when is_integer(value) and value > 0, do: {:ok, value}
  defp positive_integer(_, error), do: {:error, error}

  defp boolean(value, _error) when is_boolean(value), do: {:ok, value}
  defp boolean(_, error), do: {:error, error}

  defp operating_system("linux"), do: {:ok, :linux}
  defp operating_system("windows"), do: {:ok, :windows}
  defp operating_system("macos"), do: {:ok, :macos}
  defp operating_system("other"), do: {:ok, :other}
  defp operating_system(_), do: {:error, :invalid_os}

  defp architecture("x86_64"), do: {:ok, :x86_64}
  defp architecture("arm64"), do: {:ok, :arm64}
  defp architecture("other"), do: {:ok, :other}
  defp architecture(_), do: {:error, :invalid_arch}

  defp execution_backends(values) when is_list(values) and values != [] do
    converted = Enum.map(values, &execution_backend/1)

    if Enum.all?(converted, &match?({:ok, _}, &1)) do
      backends = Enum.map(converted, fn {:ok, value} -> value end)

      if MapSet.size(MapSet.new(backends)) == length(backends),
        do: {:ok, backends},
        else: {:error, :invalid_execution_backends}
    else
      {:error, :invalid_execution_backends}
    end
  end

  defp execution_backends(_), do: {:error, :invalid_execution_backends}

  defp execution_backend("container"), do: {:ok, :container}
  defp execution_backend("native_process"), do: {:ok, :native_process}
  defp execution_backend("virtual_machine"), do: {:ok, :virtual_machine}
  defp execution_backend(_), do: :error

  defp command_ack_status("accepted"), do: {:ok, :accepted}
  defp command_ack_status("duplicate"), do: {:ok, :duplicate}
  defp command_ack_status("rejected"), do: {:ok, :rejected}
  defp command_ack_status(_), do: {:error, :invalid_ack_status}

  defp placement_state("accepted"), do: {:ok, :accepted}
  defp placement_state("starting"), do: {:ok, :starting}
  defp placement_state("observed"), do: {:ok, :observed}
  defp placement_state("running"), do: {:ok, :running}
  defp placement_state("finished"), do: {:ok, :finished}
  defp placement_state("failed"), do: {:ok, :failed}
  defp placement_state("cancelled"), do: {:ok, :cancelled}
  defp placement_state(_), do: {:error, :invalid_placement_state}

  defp detail_code(:null), do: {:ok, nil}
  defp detail_code(nil), do: {:ok, nil}
  defp detail_code(value), do: identifier(value, :invalid_detail_code)

  defp response_code(nil), do: {:ok, nil}

  defp response_code(code) when is_atom(code),
    do: identifier(Atom.to_string(code), :invalid_response_code)

  defp response_code(_), do: {:error, :invalid_response_code}

  defp response_command(nil, _now_unix_ms), do: {:ok, :null}

  defp response_command(%NodeCommand{} = command, now_unix_ms),
    do: NodeCommand.wire_map(command, now_unix_ms)

  defp response_command(_, _now_unix_ms), do: {:error, :invalid_response_command}

  defp agent_version(value) when is_binary(value) and byte_size(value) in 1..64 do
    if String.to_charlist(value)
       |> Enum.all?(fn char ->
         char in ?0..?9 or char in ?A..?Z or char in ?a..?z or char in ~c".-+"
       end),
       do: {:ok, value},
       else: {:error, :invalid_agent_version}
  end

  defp agent_version(_), do: {:error, :invalid_agent_version}

  defp decode_json(binary) do
    try do
      case :json.decode(binary) do
        value when is_map(value) -> {:ok, value}
        _ -> {:error, :invalid_json_object}
      end
    rescue
      _ -> {:error, :invalid_json}
    end
  end
end
